tools/tdx-measure/tdvf/tdvf.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 Edgeless Systems GmbH
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package tdvf
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/edgelesssys/contrast/tdx-measure/mrtd"
+)
+
+var (
+	tableFooterGuid       = []byte{0xde, 0x82, 0xb5, 0x96, 0xb2, 0x1f, 0xf7, 0x45, 0xba, 0xea, 0xa3, 0x66, 0xc5, 0x5a, 0x08, 0x2d}
+	tdxMetadataOffsetGuid = []byte{0x35, 0x65, 0x7a, 0xe4, 0x4a, 0x98, 0x98, 0x47, 0x86, 0x5e, 0x46, 0x85, 0xa7, 0xbf, 0x8e, 0xc2}
+	metadataSignature     = "TDVF"
+)
+
+func findTdxMetadataOffset(firmare []byte) (uint32, error) {
+	footerGuid := firmare[len(firmare)-0x30:][:0x10]
+	if !bytes.Equal(footerGuid, tableFooterGuid) {
+		return 0, errors.New("can't find table footer GUID")
+	}
+
+	offset := len(firmare) - 0x32
+	tableLength := binary.LittleEndian.Uint16(firmare[offset:][:2])
+	endOffset := len(firmare) - 0x20 - int(tableLength)
+
+	for endOffset < offset {
+		entryBlockGuid := firmare[offset-0x10:][:0x10]
+		entryBlockLength := binary.LittleEndian.Uint16(firmare[offset-0x12:][:2])
+
+		if bytes.Equal(entryBlockGuid, tdxMetadataOffsetGuid) {
+			tdxOffset := binary.LittleEndian.Uint32(firmare[offset-0x16:][:4])
+			return tdxOffset, nil
+		}
+
+		offset -= int(entryBlockLength)
+	}
+
+	return 0, errors.New("can't find TDX metadata offset block entry")
+}
+
+type tdvfSection struct {
+	DataOffset     uint32
+	RawDataSize    uint32
+	MemoryAddress  uint64
+	MemoryDataSize uint64
+	Type           tdvfSectionType
+	Attributes     tdvfSectionAttributes
+}
+
+type tdvfSectionType uint32
+
+const (
+	Bbv tdvfSectionType = iota
+	Cfv
+	TdHob
+	TempMem
+	PermMem
+	Payload
+	PayloadParam
+	TdInfo
+)
+
+type tdvfSectionAttributes uint32
+
+const (
+	MrExtend tdvfSectionAttributes = 1 << iota
+	PageAug
+)
+
+func parseTdvfSections(firmare []byte) ([]tdvfSection, error) {
+	offset, err := findTdxMetadataOffset(firmare)
+	if err != nil {
+		return nil, fmt.Errorf("can't locate TDX firmware metadata offset: %w", err)
+	}
+
+	metadata := firmare[len(firmare)-int(offset):]
+	metadataHeader := metadata[:16]
+
+	if string(metadataHeader[:4]) != metadataSignature {
+		return nil, errors.New("unexpected signature")
+	}
+
+	length := binary.LittleEndian.Uint32(metadataHeader[4:][:4])
+	_ = length
```

Comment:
Comment why this is needed?

Reply:
+1

Reply:
It's not needed :D
https://github.com/edgelesssys/contrast/blob/af91458e5c6a7ed2a75b8edd2f5ef468ad740ca0/tools/tdx-measure/tdvf/tdvf.go#L83-L85

I just put that there for completeness sake should we ever need it for some reason.

Reply:
It would help if it said why it's not needed in the docstring, though. Is it because the information is redundant with the number of entries 8 bytes further down?